src/material.ts

```diff
@@ -62,316 +63,341 @@ export const defaultMaterialConfig = {
     },
 };
 
-interface Property {
+export interface Property {
     name: string;
     value: string;
 }
 
-interface MaterialSchemaJSON extends MaterialSchema, AnyObject {}
+export interface MaterialSchemaJSON extends MaterialSchema, AnyObject {}
 
-interface CheckResult {
+export interface CheckResult {
     key: string;
     name: string;
     severity: string;
     message: string;
 }
 
-export abstract class Material extends HasConsistencyChecksHasMetadataNamedDefaultableInMemoryEntity {
-    abstract src: {
-        extension: string;
-        text: string;
-    };
-
-    declare _json: MaterialSchemaJSON;
+type MaterialBaseEntity = InstanceType<
+    typeof HasConsistencyChecksHasMetadataNamedDefaultableInMemoryEntity
+>;
+
+export type MaterialBaseEntityConstructor<T extends MaterialBaseEntity = MaterialBaseEntity> = new (
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    ...args: any[]
+) => T;
+
+export function MaterialMixin<
+    T extends MaterialBaseEntityConstructor = MaterialBaseEntityConstructor,
+>(superclass: T) {
+    class MadeMaterial extends superclass {
+        declare _json: MaterialSchemaJSON;
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        constructor(...config: any[]) {
+            super(...config);
+            this.name = super.name || this.formula;
+        }
 
-    constructor(config: MaterialSchemaJSON) {
-        super(config);
-        this.name = super.name || this.formula;
-    }
+        toJSON(): MaterialJSON {
+            return {
+                ...super.toJSON(),
+                lattice: this.Lattice.toJSON(),
+                basis: this.Basis.toJSON(),
+                name: this.name,
+                isNonPeriodic: this.isNonPeriodic,
+            };
+        }
 
-    toJSON(): MaterialJSON {
-        return {
-            ...super.toJSON(),
-            lattice: this.Lattice.toJSON(),
-            basis: this.Basis.toJSON(),
-            name: this.name,
-            isNonPeriodic: this.isNonPeriodic,
-        };
-    }
+        static get defaultConfig() {
+            return defaultMaterialConfig;
+        }
 
-    static get defaultConfig() {
-        return defaultMaterialConfig;
-    }
+        get src() {
+            return this.prop<FileSourceSchema>("src");
+        }
 
-    updateFormula() {
-        this.setProp("formula", this.Basis.formula);
-        this.setProp("unitCellFormula", this.Basis.unitCellFormula);
-    }
+        set src(src: FileSourceSchema) {
+            this.setProp("src", src);
+        }
```

Comment:
These dont seem to have existed in previous Material class. Why are they added in this task?

Reply:
?

Reply:
There was 
```
abstract src: {
        extension: string;
        text: string;
    };
```
But this was not 100% correct, because we are using `src` here in `getAsPOSCAR`, though we didn't define how to fetch it